File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> current;
        dfs(0, current, 0, candidates, target, result);
        return result;
    }

private:
    void dfs(int index, vector<int>& current, int total, const vector<int>& candidates, int target, vector<vector<int>>& result) {
        if (total == target) {
            result.push_back(current);
            return;
        }
        if (index >= candidates.size() || total > target) {
            return;
        }

        
        current.push_back(candidates[index]);
        dfs(index, current, total + candidates[index], candidates, target, result);

        
        current.pop_back();
        dfs(index + 1, current, total, candidates, target, result);
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        
        def backtrack(i, cur, total):
            # base case
            if total == target:
                res.append(cur.copy())
                return
                
            # 2nd base case (out of index or sum)
            if i >= len(candidates) or total > target:
                return

            # 1st decision to include same number
            cur.append(candidates[i])
            backtrack(i, cur, total + candidates[i])
            
            # cleaning up
            cur.pop()
            # 2nd decision (to include other num)
            backtrack(i + 1, cur, total)
            
        backtrack(0, [], 0)
        return res
            